### PR TITLE
Decode numbers using json.Decoder

### PR DESCRIPTION
**Pull request by matthew1001**
If the `params` passed in for a contract deploy include very large numbers, `json.Unmarshal` to an `interface{}` doesn't work and results in an incorrect number being passed in to the contract's constructor.

To recreate the issue, if you checkout just the UT changes from this branch and run them it will fail.

This PR switches `json.Unmarshal` to `json.Decode`.

### CODE REVIEW

File: internal/ethereum/deploy_contract_prepare.go [hunk 1]
```diff
@@ -95,7 +96,9 @@ func (c *ethConnector) prepareDeployData(ctx context.Context, req *ffcapi.Contra
 	ethParams := make([]interface{}, len(req.Params))
 	for i, p := range req.Params {
 		if p != nil {
-			err := json.Unmarshal([]byte(*p), &ethParams[i])
+			decodedParam := json.NewDecoder(bytes.NewReader([]byte(*p)))
+			decodedParam.UseNumber()
```
Comment (EnriqueL8) [inline internal/ethereum/deploy_contract_prepare.go:100]: From what I can read this stores the number in the string representation, have you checked that FireFly Signer supports this?

Comment (EnriqueL8) [reply internal/ethereum/deploy_contract_prepare.go:100]: Just thinking of how the RLP encode code there would treat this

**Review by EnriqueL8 — COMMENTED** [review]
Looks good - just one comment on testing with FireFly Signer and understanding if not only the number is there in RLP but also the things around it like where the number starts

File: internal/ethereum/deploy_contract_prepare.go (hunk 1 above)
Comment (EnriqueL8) [reply internal/ethereum/deploy_contract_prepare.go:100]: So it feels like it was just going wrong before it hit FireFly Signer https://github.com/hyperledger/firefly-signer/blob/5256116539ed83a23a52103f6260f4887abe0e85/pkg/abi/abiencode.go#L181

Comment (Chengxuan) [reply internal/ethereum/deploy_contract_prepare.go:100]: I suspect we need to fix the issue in https://github.com/hyperledger/firefly-common/blob/main/pkg/fftypes/jsonany.go#L51

File: internal/ethereum/deploy_contract_prepare.go [hunk 2]
```diff
@@ -95,7 +96,9 @@ func (c *ethConnector) prepareDeployData(ctx context.Context, req *ffcapi.Contra
...
+			err := decodedParam.Decode(&ethParams[i])
```
Comment (Chengxuan) [inline internal/ethereum/deploy_contract_prepare.go:101]: I think we need to fix the issue in https://github.com/hyperledger/firefly-common/blob/main/pkg/fftypes/jsonany.go#L51 if the issue is how data gets unmarshalled

Comment (matthew1001) [reply internal/ethereum/deploy_contract_prepare.go:101]: Agreed - and it probably needs a quick search of the code for other places where `json.Unmarshal()` is used as well

Comment (matthew1001) [reply internal/ethereum/deploy_contract_prepare.go:101]: Going to prove out this fix with an end-to-end contract deploy first to check that the fix is good (beyond the UTs)

File: internal/ethereum/deploy_contract_prepare.go (hunk 1 above)
Comment (peterbroadhurst) [reply internal/ethereum/deploy_contract_prepare.go:100]: Trying to understand this. The code before seemed correct:
```go
// Parse the params into the standard semantics of Go JSON unmarshalling, with []interface{}
```

Comment (peterbroadhurst) [reply internal/ethereum/deploy_contract_prepare.go:100]: The intent of this code should be to take:
- take raw `json.RawMessage` etc still input from `req.Params`
- for each, do a standard `json.Unmarshal()` to `interface{}` so Golang picks the type
- pass that down to `decodedParam.Decode` to do type-specific processing on

File: internal/ethereum/deploy_contract_prepare.go (hunk 2 above)
Comment (peterbroadhurst) [reply internal/ethereum/deploy_contract_prepare.go:101]: Sorry @matthew1001 - I'm not sure I understand why there's a concern with `json.Unmarshal()`

File: internal/ethereum/deploy_contract_prepare_test.go [hunk 3]
```diff
@@ -54,6 +55,37 @@ const samplePrepareDeployTX = `{
...
+		    {
+			    "internalType":" string",
+			    "name": "x",
+			    "type": "string"
+		    }
+		],
+		"outputs":[],
+		"type":"constructor"
+	}],
+	"params": [ 10000000000000000000000001, "some-text" ]
```
Comment (peterbroadhurst) [inline internal/ethereum/deploy_contract_prepare_test.go:86]: This input is broken in JSON, it's attempting to use a number that does not fit into int64

**Review by peterbroadhurst — CHANGES_REQUESTED** [review]
...

File: internal/ethereum/deploy_contract_prepare.go (hunk 2 above)
Comment (peterbroadhurst) [reply internal/ethereum/deploy_contract_prepare.go:101]: Ok - I do think I now understand, but the right fix (which I think would probably include `json.Unmarshal`) would take time to investigate

**Review by peterbroadhurst — COMMENTED** [review]
ok - sorry, after some misunderstanding I now understand what this PR is trying to achieve.

This PR is an enhancement to support:
- JSON input in the `params` array is able to contain JSON `number` elements larger than fit into an `int64`.

One part of the fix, is that `Golang` provides an option on `json.Decoder` (not available on `json.Unmarshal`) to decode `number`s into `json.Number` types (which are `strings`).

If that is done in this one piece of the stack, then it has implications for other places that need to be reviewed:
- `firefly`
- `firefly-signer`

With this done, a user choosing to generate a JSON input containing JSON `number` > that is possible to generate in Javascript/Java/Go serialization of numbers could be supported by the stack.

**Comment by matthew1001** [comment]
Thanks for the comments @peterbroadhurst, @Chengxuan, @EnriqueL8 - I might grab some in-person time with @EnriqueL8 to discuss options & other places in the stack that will be affected. I'll circle back round on this PR once I've done that.

**Pull request closed without merging**